js/packages/cli/src/commands/upload.ts

```diff
@@ -1,211 +1,166 @@
 import { chunks } from '../helpers/various';
+import { AssetKey } from '../types';
 
-export async function upload(
-  files: string[],
-  cacheName: string,
-  env: string,
-  keypair: string,
-  totalNFTs: number,
-  storage: string,
-  retainAuthority: boolean,
-  mutable: boolean,
-  rpcUrl: string,
-  ipfsCredentials: ipfsCreds,
-  awsS3Bucket: string,
-  batchSize: number,
-): Promise<boolean> {
-  let uploadSuccessful = true;
-
-  const savedContent = loadCache(cacheName, env);
-  const cacheContent = savedContent || {};
-
-  if (!cacheContent.program) {
-    cacheContent.program = {};
-  }
-
-  let existingInCache = [];
-  if (!cacheContent.items) {
-    cacheContent.items = {};
-  } else {
-    existingInCache = Object.keys(cacheContent.items);
-  }
-
-  const seen = {};
-  const newFiles = [];
-  const EXTENSION_ARRAY = [EXTENSION_PNG, EXTENSION_GIF, EXTENSION_JPG];
+/**
+ * The Cache object, represented in its minimal form.
+ */
+type Cache = {
+  program: {
+    config?: string;
+  };
+  items: {
+    [key: string]: any;
+  };
+};
 
-  files.forEach(f => {
-    const extensionRegex = new RegExp(
-      `/(${EXTENSION_PNG})|(${EXTENSION_GIF})|(${EXTENSION_JPG})/g`,
-    );
-    if (!seen[f.replace(extensionRegex, '').split('/').pop()]) {
-      seen[f.replace(extensionRegex, '').split('/').pop()] = true;
-      newFiles.push(f);
-    }
-  });
-  existingInCache.forEach(f => {
-    if (!seen[f]) {
-      seen[f] = true;
-      const imageLink = cacheContent?.items?.[f]?.imageLink;
-      if (imageLink && EXTENSION_ARRAY.includes(path.extname(imageLink))) {
-        newFiles.push(f + path.extname(imageLink));
-      } else {
-        log.error(`Unknown file extension or image (${f}): ${imageLink}`);
-        newFiles.push(f + '.png');
+/**
+ * The Manifest object for a given asset.
+ * This object holds the contents of the asset's JSON file.
+ * Represented here in its minimal form.
+ */
+type Manifest = {
+  image: string;
+  name: string;
+  symbol: string;
+  seller_fee_basis_points: number;
+  properties: {
+    files: Array<{ type: string; uri: string }>;
+    creators: Array<{
+      address: string;
+      share: number;
+    }>;
+  };
+};
+/**
+ * From the Cache object & a list of file paths, return a list of asset keys
+ * (filenames without extension nor path) that should be uploaded, sorted numerically in ascending order.
+ * Assets which should be uploaded either are not present in the Cache object,
+ * or do not truthy value for the `link` property.
+ */
+function getAssetKeysNeedingUpload(
+  items: Cache['items'],
+  files: string[],
+): AssetKey[] {
+  const all = [
+    ...new Set([
+      ...Object.keys(items),
+      ...files.map(filePath => path.basename(filePath)),
+    ]),
+  ];
+  const keyMap = {};
+  return all
+    .filter(k => !k.includes('.json'))
+    .reduce((acc, assetKey) => {
+      const ext = path.extname(assetKey);
+      const key = path.basename(assetKey, ext);
```

Comment:
`if (!imageExtensionArray.includes(ext)) throw new Error` or something like that